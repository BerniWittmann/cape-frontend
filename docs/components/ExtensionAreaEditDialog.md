# ExtensionAreaEditDialog

A Dialog to edit an Extension Area

