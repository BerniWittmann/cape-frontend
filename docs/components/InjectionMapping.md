# InjectionMapping

A Component to edit an Injection Mapping

## Props

<!-- @vuese:InjectionMapping:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|injectionMapping|the Injection Mapping|`Object`|`true`|-|

<!-- @vuese:InjectionMapping:props:end -->


