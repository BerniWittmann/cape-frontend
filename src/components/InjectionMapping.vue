<template>
  <p>TODO</p>
</template>

<script>

/*
 * @vuese
 * @group Components
 *
 * A Component to edit an Injection Mapping
 */
export default {
  name: 'InjectionMapping',
  props: {
    // the Injection Mapping
    injectionMapping: {
      type: Object,
      required: true
    }
  }
}

</script>

<style lang="scss">

</style>
