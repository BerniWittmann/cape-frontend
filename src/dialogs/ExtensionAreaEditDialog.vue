<template>
  <el-dialog :title="title" :visible.sync="showExtensionArea" width="70%">
    <el-collapse v-model="activeInjection" accordion>
      <el-collapse-item v-for="injectionMapping in injectionMappings" :key="injectionMapping.id" :name="injectionMapping.id || 'new'">
        <template slot="title">
          {{ getTitlePart(injectionMapping, 'contextSituation') }} <i class="title-icon el-icon-back"></i> {{ getTitlePart(injectionMapping, 'injectedProcess') }}
        </template>
        <injection-mapping  :injection-mapping="injectionMapping"></injection-mapping>
      </el-collapse-item>
    </el-collapse>
    <el-button plain icon="el-icon-plus" size="small" @click="addInjectionMapping()" id="add" v-if="noExistingNewInjection">
      {{ $t('injection_mapping.add') }}
    </el-button>
  </el-dialog>
</template>

<script>

import InjectionMapping from '@/components/InjectionMapping.vue'
import InjectionMappingModel from '@/models/injectionMapping.js'

/*
 * @vuese
 * @group Dialogs
 *
 * A Dialog to edit an Extension Area
 */
export default {
  name: 'ExtensionAreaEditDialog',
  components: {
    InjectionMapping
  },

  data() {
    return {
      activeInjection: ''
    }
  },

  computed: {
    activeProcess() {
      return this.$store.state.process.activeProcess
    },
    showExtensionArea: {
      get: function () {
        return !!this.activeProcess
      },
      set: function () {
        this.closeDialog()
      }
    },

    title() {
      return this.$route.params.title || this.$t('extension_area.title')
    },

    injectionMappings() {
      return this.$store.getters['injectionMapping/getInjectionMappings'](this.$route.params.processID, this.$route.params.extensionAreaID)
    },
    noExistingNewInjection() {
      return !(this.injectionMappings.filter(a => a.id === undefined).length > 0)
    }
  },

  methods: {
    closeDialog() {
      this.$router.back()
    },

    getTitlePart(injectionMapping, name) {
      if (!injectionMapping[name]) return this.$t('injection_mapping.undefined_' + name)
      return injectionMapping[name].name
    },
    addInjectionMapping() {
      this.$store.dispatch('injectionMapping/add', new InjectionMappingModel({ 'extension_area_id': this.$route.params.extensionAreaID, 'process_id': this.$route.params.processID })).then(() => {
        this.activeInjection = 'new'
      })
    }
  }
}
</script>

<style scoped lang="scss">
.title-icon {
  transform: rotate(180deg);
  margin-left: 10px;
  margin-right: 10px;
}
#add{
  margin-bottom: 10px;
  margin-top: 10px;
}
</style>
