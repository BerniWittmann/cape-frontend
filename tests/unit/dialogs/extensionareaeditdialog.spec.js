import { shallowMount } from '@vue/test-utils'
import { i18n } from '../setupPlugins'

import moment from 'moment'
import { Button, Dialog } from 'element-ui'

import ExtensionAreaEditDialog from '@/dialogs/ExtensionAreaEditDialog.vue'
import InjectionMapping from '@/components/InjectionMapping'

describe('Dialogs', () => {
  describe('Extension Area Edit Dialog', () => {
    let store
    let cmp

    const date = moment(moment().utc())
    const router = {
      back: jest.fn(),
      push: jest.fn()
    }
    const route = {
      params: {
        processID: 42,
        extensionAreaID: 'EA_1',
        title: 'My Title'
      }
    }
    let injections = []
    beforeEach(() => {
      injections = [{
        id: '1',
        processID: '1',
        extensionAreaID: 'EA_1',
        foo: 'bar',
        injectedProcess: {
          id: 'p1',
          name: 'Process 1'
        },
        contextSituation: {
          id: 'cs1',
          name: 'Context Situation 1'
        }
      }, {
        id: '2',
        processID: '1',
        extensionAreaID: 'EA_1',
        foo: 'bar2',
        injectedProcess: {
          id: 'p2',
          name: 'Process 2'
        },
        contextSituation: {
          id: 'cs2',
          name: 'Context Situation 2'
        }
      }]
      store = {
        state: {
          process: {
            activeProcess: {
              id: '1',
              name: 'My Process',
              createdAt: date.clone().subtract(14, 'days'),
              lastEditedAt: date.clone().subtract(2, 'days'),
              svg: 'text height="30" more text="even more Text"',
              description: 'My Description',
              xml: '<?xml version="1.0" encoding="UTF-8"?><bpmn:Process>Test Process</bpmn:Process>',
              tags: [{
                id: '42',
                name: 'First Tag',
                color: '#FF0000'
              }, {
                id: '43',
                name: 'Second Tag',
                color: '#FFFF00'
              }]
            }
          }
        },
        getters: {
          'injectionMapping/getInjectionMappings': jest.fn().mockReturnValue(injections)
        }
      }
      render()
    })

    function render() {
      cmp = shallowMount(ExtensionAreaEditDialog, {
        i18n,
        mocks: {
          $store: store,
          $router: router,
          $route: route
        },
        stubs: {
          ElButton: Button,
          ElDialog: Dialog,
          ElCollapseItem: '<div class="el-collapse-item"><div class="title"><slot name="title"></slot></div><div class="content"><slot></slot></div></div>'
        }
      })
    }

    it('can render', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('can be closed', () => {
      expect(cmp.html()).toMatchSnapshot()
      cmp.find('.el-dialog__header button').trigger('click')
      expect(router.back).toHaveBeenCalled()
    })

    it('displays the title from the route params', () => {
      expect(cmp.find('.el-dialog__title').text()).toEqual('My Title')
    })

    it('displays a fallback if title is not given via route params', () => {
      route.params.title = undefined
      render()
      expect(cmp.html()).toMatchSnapshot()
      expect(cmp.find('.el-dialog__title').text()).toEqual('extension_area.title')
    })

    it('renders the injection mappings', () => {
      expect(store.getters['injectionMapping/getInjectionMappings']).toHaveBeenCalledWith(42, 'EA_1')
      const ims = cmp.findAll(InjectionMapping)
      expect(ims.length).toEqual(2)
      expect(ims.at(0).props('injectionMapping')).toEqual({
        id: '1',
        processID: '1',
        extensionAreaID: 'EA_1',
        foo: 'bar',
        injectedProcess: {
          id: 'p1',
          name: 'Process 1'
        },
        contextSituation: {
          id: 'cs1',
          name: 'Context Situation 1'
        }
      })
      expect(ims.at(1).props('injectionMapping')).toEqual({
        id: '2',
        processID: '1',
        extensionAreaID: 'EA_1',
        foo: 'bar2',
        injectedProcess: {
          id: 'p2',
          name: 'Process 2'
        },
        contextSituation: {
          id: 'cs2',
          name: 'Context Situation 2'
        }
      })
    })

    it('renders a title for the injection mappings', (done) => {
      let items = cmp.findAll('.el-collapse-item')
      expect(items.length).toEqual(2)
      expect(items.at(0).find('.title').text()).toEqual('Context Situation 1  Process 1')
      expect(items.at(1).find('.title').text()).toEqual('Context Situation 2  Process 2')

      injections.push({
        id: '3',
        processID: '1',
        extensionAreaID: 'EA_1',
        foo: 'bar2',
        injectedProcess: undefined,
        contextSituation: undefined
      })
      injections.push({
        id: '4',
        processID: '1',
        extensionAreaID: 'EA_1',
        foo: 'bar2',
        injectedProcess: {
          id: 'p2',
          name: 'Process 2'
        },
        contextSituation: undefined
      })
      injections.push({
        id: '5',
        processID: '1',
        extensionAreaID: 'EA_1',
        foo: 'bar2',
        injectedProcess: undefined,
        contextSituation: {
          id: 'cs2',
          name: 'Context Situation 2'
        }
      })
      render()

      cmp.vm.$nextTick(() => {
        items = cmp.findAll('.el-collapse-item')
        expect(items.length).toEqual(5)
        expect(items.at(2).find('.title').text()).toEqual('injection_mapping.undefined_contextSituation  injection_mapping.undefined_injectedProcess')
        expect(items.at(3).find('.title').text()).toEqual('injection_mapping.undefined_contextSituation  Process 2')
        expect(items.at(4).find('.title').text()).toEqual('Context Situation 2  injection_mapping.undefined_injectedProcess')
        done()
      })
    })
  })
})
