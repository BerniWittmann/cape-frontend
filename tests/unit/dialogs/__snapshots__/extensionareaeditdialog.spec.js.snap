// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialogs Extension Area Edit Dialog can be closed 1`] = `
<div class="el-dialog__wrapper" name="dialog-fade" style="z-index: 2003;">
  <div role="dialog" aria-modal="true" aria-label="My Title" class="el-dialog" style="margin-top: 15vh; width: 70%;">
    <div class="el-dialog__header"><span class="el-dialog__title">My Title</span><button type="button" aria-label="Close" class="el-dialog__headerbtn"><i class="el-dialog__close el-icon el-icon-close"></i></button></div>
    <div class="el-dialog__body">
      <elcollapse-stub accordion="true" value="">
        <div class="el-collapse-item" name="1">
          <div class="title">
            Context Situation 1 <i class="title-icon el-icon-back"></i> Process 1
          </div>
          <div class="content">
            <injectionmapping-stub injectionmapping="[object Object]"></injectionmapping-stub>
          </div>
        </div>
        <div class="el-collapse-item" name="2">
          <div class="title">
            Context Situation 2 <i class="title-icon el-icon-back"></i> Process 2
          </div>
          <div class="content">
            <injectionmapping-stub injectionmapping="[object Object]"></injectionmapping-stub>
          </div>
        </div>
      </elcollapse-stub>
    </div>
    <!---->
  </div>
</div>
`;

exports[`Dialogs Extension Area Edit Dialog can render 1`] = `
<div class="el-dialog__wrapper" name="dialog-fade" style="z-index: 2001;">
  <div role="dialog" aria-modal="true" aria-label="My Title" class="el-dialog" style="margin-top: 15vh; width: 70%;">
    <div class="el-dialog__header"><span class="el-dialog__title">My Title</span><button type="button" aria-label="Close" class="el-dialog__headerbtn"><i class="el-dialog__close el-icon el-icon-close"></i></button></div>
    <div class="el-dialog__body">
      <elcollapse-stub accordion="true" value="">
        <div class="el-collapse-item" name="1">
          <div class="title">
            Context Situation 1 <i class="title-icon el-icon-back"></i> Process 1
          </div>
          <div class="content">
            <injectionmapping-stub injectionmapping="[object Object]"></injectionmapping-stub>
          </div>
        </div>
        <div class="el-collapse-item" name="2">
          <div class="title">
            Context Situation 2 <i class="title-icon el-icon-back"></i> Process 2
          </div>
          <div class="content">
            <injectionmapping-stub injectionmapping="[object Object]"></injectionmapping-stub>
          </div>
        </div>
      </elcollapse-stub>
    </div>
    <!---->
  </div>
</div>
`;

exports[`Dialogs Extension Area Edit Dialog displays a fallback if title is not given via route params 1`] = `
<div class="el-dialog__wrapper" name="dialog-fade" style="z-index: 2009;">
  <div role="dialog" aria-modal="true" aria-label="extension_area.title" class="el-dialog" style="margin-top: 15vh; width: 70%;">
    <div class="el-dialog__header"><span class="el-dialog__title">extension_area.title</span><button type="button" aria-label="Close" class="el-dialog__headerbtn"><i class="el-dialog__close el-icon el-icon-close"></i></button></div>
    <!---->
    <!---->
  </div>
</div>
`;
